internal/commands/logs/list.go

```diff
@@ -38,25 +44,79 @@ func (cmd *CommandList) Inputs() cli.InputResolver {
 
 // Handler is the command handler
 func (cmd *CommandList) Handler(profile *cli.Profile, ui terminal.UI, clients cli.Clients) error {
+	cmdStart := time.Now() // for use with tail later
+
 	app, err := cli.ResolveApp(ui, clients.Realm, cmd.inputs.Filter())
 	if err != nil {
 		return err
 	}
 
-	logs, err := clients.Realm.Logs(app.GroupID, app.ID, realm.LogsOptions{
+	opts := realm.LogsOptions{
 		Types:      cmd.inputs.Types,
 		ErrorsOnly: cmd.inputs.Errors,
-		Start:      cmd.inputs.Start.Time,
-		End:        cmd.inputs.End.Time,
-	})
+	}
+	if !cmd.inputs.Tail {
+		opts.Start = cmd.inputs.Start.Time
+		opts.End = cmd.inputs.End.Time
+	}
+
+	logs, err := clients.Realm.Logs(app.GroupID, app.ID, opts)
 	if err != nil {
 		return err
 	}
 
-	sort.SliceStable(logs, func(i, j int) bool {
-		return logs[i].Started.Before(logs[j].Started)
-	})
+	if cmd.inputs.Tail && len(logs) > tailLookBehind {
+		logs = logs[0:tailLookBehind]
+	}
+
+	printLogs(ui, logs)
+	if !cmd.inputs.Tail {
+		return nil // if not tailing, command stops here
+	}
+
+	logsCh, errCh, closeCh := make(chan realm.Logs), make(chan error, 1), make(chan struct{}, 1)
+
+	opts.Start = cmdStart
+	go pollForLogs(clients.Realm, app.GroupID, app.ID, opts, logsCh, errCh, closeCh)
+
+	for {
+		select {
+		case logs := <-logsCh:
+			printLogs(ui, logs)
+		case err := <-errCh:
+			return err
+		case <-cmd.inputs.sigShutdown:
+			close(closeCh)
+			return nil
+		}
+	}
+	return nil
+}
+
+func pollForLogs(realmClient realm.Client, groupID, appID string, opts realm.LogsOptions, logsCh chan<- realm.Logs, errCh chan<- error, closeCh <-chan struct{}) {
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-closeCh:
+			return
+		case <-ticker.C:
+			startNext := time.Now()
 
+			logs, err := realmClient.Logs(groupID, appID, opts)
+			if err != nil {
+				errCh <- err
+			}
+			logsCh <- logs
```

Comment:
seems like we could have a potential race between taking from `logsCh` and `errCh` in the `select...for` loop on line 82, but maybe that's okay (since if `logsCh` happens to take first it'll just loop back to taking from `errCh` and then exit). just confirming the intended behavior here.

or potentially... if the `errCh` take happens first, we exit, and then nothing is pulling from `logsCh`, so this goroutine will block forever. you might need to make sure you drain `logsCh` in that case. or potentially just return in the error case to exit.

in reality this probably doesn't matter, since the whole program will shut down and exit, but I still think it's good to get in the habit of making sure we're always managing goroutine exit conditions properly.

Reply:
discussed offline, adding a `return` after sending on the `errCh`, and double-checking the `closeCh` before sending on `logsCh` to prevent this goroutine from hanging